app/Http/Controllers/DocuSignController.php

```diff
@@ -97,71 +142,273 @@ public function signAgreement(Request $request)
 
-                config(
-                    'docusign.membership_agreement.ingest_mailbox_name'
-                ).'_UserName' => config('app.name'),
+        /** @var \DocuSign\eSign\Client\Auth\OAuthToken $tokens */
+        $tokens = $docusign->generateAccessToken(
+            client_id: config('docusign.client_id'),
+            client_secret: config('docusign.client_secret'),
+            code: $request->code
+        )[0];
 
-                config(
-                    'docusign.membership_agreement.ingest_mailbox_name'
-                ).'_Email' => config('docusign.ingest_mailbox'),
+        /** @var \DocuSign\eSign\Client\Auth\UserInfo $userinfo */
+        $userinfo = $docusign->getUserInfo($tokens->getAccessToken())[0];
 
-                config(
-                    'docusign.membership_agreement.archive_mailbox_name'
-                ).'_UserName' => 'Membership Agreement Archives',
+        $userInSameAccount = false;
 
-                config(
-                    'docusign.membership_agreement.archive_mailbox_name'
-                ).'_Email' => config('services.membership_agreement_archive_email'),
-            ]
-        );
+        /** @var \DocuSign\eSign\Client\Auth\Account $account */
+        foreach ($userinfo->getAccounts() as $account) {
+            if ($account->getAccountId() === config('docusign.account_id')) {
+                $userInSameAccount = true;
+            }
+        }
+
+        if (! $userInSameAccount) {
+            $userinfo_serialized = DocuSign::serializeUserInfo($userinfo);
+
+            Log::info($userinfo_serialized);
+
+            abort(401);
+        }
+
+        switch ($request->session()->get('next')) {
+            case 'storeGlobalTokens':
+                DocuSignToken::storeTokens($tokens);
+
+                $userinfo_serialized = DocuSign::serializeUserInfo($userinfo);
+
+                Log::info($userinfo_serialized);
+
+                return response()->json($userinfo_serialized);
+            case 'signAgreement':
+                /** @var \App\Models\User $user */
+                $user = $request->user();
+
+                if ($userinfo->getEmail() !== $user->uid.'@gatech.edu') {
+                    $userinfo_serialized = DocuSign::serializeUserInfo($userinfo);
+
+                    Log::info($userinfo_serialized);
+
+                    abort(401);
+                }
+
+                $user->docusign_access_token = $tokens->getAccessToken();
+                $user->docusign_access_token_expires_at = Carbon::now()->addSeconds($tokens->getExpiresIn() - 60);
+                $user->docusign_refresh_token = $tokens->getRefreshToken();
+                $user->docusign_refresh_token_expires_at = Carbon::now()->addDays(29);
+                $user->save();
+
+                return $this->signAgreement($request);
+        }
+    }
+
+    public function complete(Request $request)
+    {
+        if (! $request->hasValidSignatureWhileIgnoring(['event'])) {
```

Comment:
The `event` parameter is not signed, so someone could potentially replay requests with different `event`s. I don't see a clean way to prevent that without immediately making a GET call to DocuSign, which the DocuSign docs specifically discourage.

Webhooks are too slow to rely on for interactive signing.